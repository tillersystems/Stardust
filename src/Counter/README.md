# Counter

### Usage

```jsx
import { Counter } from '@tillersystems/stardust';
```

<!-- STORY -->

### Properties

- `appearance` - Button appareance.
- `className` - className needed by styled component.
- `max` - Maximum value allowed.
- `min` - Minimum value allowed.
- `onIncrement` - Callback function called on increment.
- `onDecrement` - Callback function called on decrement.
- `step` - step for increment/decrement value.
- `value` - incremented/decremented value. The value is set to 0 if no value provided.
- `width` - the width of the input.

| propName      | propType | defaultValue | isRequired |
| ------------- | :------: | :----------: | :--------: |
| `appareance`  | `string` | `secondary`  |     -      |
| `className`   | `string` |     `''`     |     -      |
| `max`         | `number` |    `100`     |     -      |
| `min`         | `number` |     `0`      |     -      |
| `onIncrement` |  `func`  |  `() => {}`  |     -      |
| `onDecrement` |  `func`  |  `() => {}`  |     -      |
| `step`        | `number` |     `1`      |     -      |
| `value`       | `number` |     `0`      |     -      |
| `width`       | `string` |    `5rem`    |     -      |

### Example

#### Standard use case

The following snippet shows how to create a increment component.

```jsx
import { Counter } from '@tillersystems/stardust';

render() {
  return (
    <Counter
        step={myStepValue}
        max={maxValue}
        min={minValue}
        onIncrement={() => {}}
        onDecrement={() => {}}
        appearance={appearance}
        width={`${widthValue}rem`}
      />
  );
}
```
