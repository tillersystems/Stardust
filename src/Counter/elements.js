import styled from 'styled-components';

export const FakeInput = styled.span`
  display: flex;
  justify-content: center;
  align-items: center;

  width: ${({ width }) => width};

  background: ${({ theme: { palette } }) => palette.white};
  border-top: 1px solid ${({ theme: { palette } }) => palette.lightGrey};
  border-bottom: 1px solid ${({ theme: { palette } }) => palette.lightGrey};

  color: ${({ theme: { palette } }) => palette.darkBlue};
`;
