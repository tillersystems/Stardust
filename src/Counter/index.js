import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import { Button } from '..';
import { FakeInput } from './elements';

const { func, number, string } = PropTypes;

/**
 * A Counter is a number value that can be updated by two Buttons surrounding the displayed value.
 *
 * @return {jsx}
 */
class Counter extends PureComponent {
  state = {
    count: this.props.value, // eslint-disable-line react/destructuring-assignment,
  };

  /**
   * Component Did Update
   * Method invoked immediately after updating occur.
   */
  componentDidUpdate(prevProps) {
    const { value } = this.props;

    // Update state only if props are changed
    if (prevProps.value !== value) {
      this.setState({
        count: value,
      });
    }
  }

  /**
   * Handle Increment click.
   */
  handleIncrementClick = () => {
    const { count } = this.state;
    const { step, max, onIncrement } = this.props;

    this.setState(
      prevState => ({
        count: count === max ? prevState.count : prevState.count + step,
      }),
      () => (count === max ? null : onIncrement()),
    );
  };

  /**
   * Handle Decrement click.
   */
  handleDecrementClick = () => {
    const { count } = this.state;
    const { step, min, onDecrement } = this.props;

    this.setState(
      prevState => ({
        count: count === min ? prevState.count : prevState.count - step,
      }),
      () => (count === min ? null : onDecrement()),
    );
  };

  /**
   * Renders the component.
   *
   * @return {jsx}
   */
  render() {
    const { className, step, max, min, width, appearance } = this.props;
    const { count } = this.state;

    return (
      <div
        role="group"
        className={className}
        tabIndex="-1"
        step={step}
        max={max}
        min={min}
        width={width}
      >
        <Button appearance={appearance} onClick={this.handleDecrementClick} data-testid="decrement">
          -
        </Button>
        <FakeInput width={width} data-testid="fakeinput">
          {count}
        </FakeInput>
        <Button appearance={appearance} onClick={this.handleIncrementClick} data-testid="increment">
          +
        </Button>
      </div>
    );
  }
}

/**
 * PropTypes Validation.
 */
Counter.propTypes = {
  /**
   * Buttons appearance prop customizing the style
   */
  appearance: string,

  /**
   * ClassName needed by styled components
   */
  className: string,

  /**
   * Maximum value allowed
   */
  max: number,

  /**
   * Minimum value allowed
   */
  min: number,

  /**
   * Callback function called on increment
   */
  onIncrement: func,

  /**
   * Callback function called on decrement
   */
  onDecrement: func,

  /**
   * Step for incrementing/decrementing value
   */
  step: number,

  /**
   * Value of counter. The value is set to 0 if no value is provided by prop
   */
  value: number,

  /**
   * The width of the input
   */
  width: string,
};

/**
 * Default props.
 */
Counter.defaultProps = {
  appearance: 'secondary',
  className: '',
  max: 1000,
  min: 0,
  onIncrement: () => {},
  onDecrement: () => {},
  step: 1,
  value: 0,
  width: '5rem',
};

export default styled(Counter)`
  display: flex;
  ${Button} {
    border-radius: 0;
    font-size: ${({ theme: { fonts } }) => fonts.size.h5};

    &:first-of-type {
      border-top-left-radius: ${({ theme: { dimensions } }) => dimensions.radius};
      border-bottom-left-radius: ${({ theme: { dimensions } }) => dimensions.radius};
    }

    &:last-of-type {
      border-top-right-radius: ${({ theme: { dimensions } }) => dimensions.radius};
      border-bottom-right-radius: ${({ theme: { dimensions } }) => dimensions.radius};
    }
  }
`;
