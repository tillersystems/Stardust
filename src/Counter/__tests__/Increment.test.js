import React from 'react';
import { fireEvent } from 'react-testing-library';

import Counter from '..';

describe('<Counter />', () => {
  test('should render without a problem', () => {
    const { container } = render(<Counter />);

    expect(container.firstChild).toMatchSnapshot();
  });

  test('should increment counter', () => {
    const { getByTestId } = render(
      <Counter step={1} max={100} min={0} value={0} appearance="secondary" width="5rem" />,
    );
    const incrementButton = getByTestId('increment');
    const fakeinputValue = getByTestId('fakeinput');

    fireEvent.click(incrementButton);

    expect(fakeinputValue).toHaveTextContent('1');
  });

  test('should decrement counter', () => {
    const { getByTestId } = render(
      <Counter step={1} max={100} min={0} value={10} appearance="secondary" width="5rem" />,
    );
    const decrementButton = getByTestId('decrement');
    const fakeinputValue = getByTestId('fakeinput');

    fireEvent.click(decrementButton);

    expect(fakeinputValue).toHaveTextContent('9');
  });

  test('should decrement counter until the min value', () => {
    const { getByTestId } = render(
      <Counter step={1} max={100} min={0} value={99} appearance="secondary" width="5rem" />,
    );
    const incrementButton = getByTestId('increment');
    const fakeinputValue = getByTestId('fakeinput');

    expect(fakeinputValue).toHaveTextContent('99');

    fireEvent.click(incrementButton);

    expect(fakeinputValue).toHaveTextContent('100');

    fireEvent.click(incrementButton);

    expect(fakeinputValue).toHaveTextContent('100');
  });

  test('should decrement counter until the min value', () => {
    const { getByTestId } = render(
      <Counter step={1} max={100} min={0} value={1} appearance="secondary" width="5rem" />,
    );
    const decrementButton = getByTestId('decrement');
    const fakeinputValue = getByTestId('fakeinput');

    expect(fakeinputValue).toHaveTextContent('1');

    fireEvent.click(decrementButton);

    expect(fakeinputValue).toHaveTextContent('0');

    fireEvent.click(decrementButton);

    expect(fakeinputValue).toHaveTextContent('0');
  });

  test('should call onIncrement callback function', () => {
    const spy = jest.fn();
    const { getByTestId } = render(
      <Counter
        step={1}
        max={100}
        min={0}
        onIncrement={spy}
        value={10}
        appearance="secondary"
        width="5rem"
      />,
    );
    const incrementButton = getByTestId('increment');

    fireEvent.click(incrementButton);

    expect(spy).toHaveBeenCalledTimes(1);
  });

  test('should call onDecrement callback function', () => {
    const spy = jest.fn();
    const { getByTestId } = render(
      <Counter
        step={1}
        max={100}
        min={0}
        onDecrement={spy}
        value={10}
        appearance="secondary"
        width="5rem"
      />,
    );
    const decrementButton = getByTestId('decrement');

    fireEvent.click(decrementButton);

    expect(spy).toHaveBeenCalledTimes(1);
  });
});
