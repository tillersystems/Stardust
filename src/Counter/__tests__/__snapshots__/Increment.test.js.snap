// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Counter /> should render without a problem 1`] = `
.c1 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  max-width: 100%;
  width: auto;
  padding: 0.75rem 1.6rem;
  border-radius: 0.4rem;
  font-size: 1.4rem;
  line-height: 1.57;
  color: hsl(0,100%,100%);
  cursor: pointer;
  outline: none;
  -webkit-transition: background 200ms ease,box-shadow 200ms ease;
  transition: background 200ms ease,box-shadow 200ms ease;
  color: hsl(207,13%,45%);
  background: linear-gradient(180deg,hsl(0,100%,100%) 0%,hsl(240,33%,99%) 100%);
  border: 1px solid hsl(207,22%,90%);
  padding: 0.75rem 1.6rem;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 5rem;
  background: hsl(0,100%,100%);
  border-top: 1px solid hsl(207,22%,90%);
  border-bottom: 1px solid hsl(207,22%,90%);
  color: hsl(213,17%,20%);
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 .c2 {
  border-radius: 0;
  font-size: 2.0rem;
}

.c0 .c2:first-of-type {
  border-top-left-radius: 0.4rem;
  border-bottom-left-radius: 0.4rem;
}

.c0 .c2:last-of-type {
  border-top-right-radius: 0.4rem;
  border-bottom-right-radius: 0.4rem;
}

<div
  class="c0"
  max="1000"
  min="0"
  role="group"
  step="1"
  tabindex="-1"
  width="5rem"
>
  <button
    class="c1 c2 "
    data-testid="decrement"
    type="button"
  >
    -
  </button>
  <span
    class="c3"
    data-testid="fakeinput"
    width="5rem"
  >
    0
  </span>
  <button
    class="c1 c2 "
    data-testid="increment"
    type="button"
  >
    +
  </button>
</div>
`;
